src/libstd/sys/unix/ext/net.rs

```diff
@@ -245,6 +304,340 @@ impl fmt::Debug for SocketAddr {
+    pub fn get_fds_by_vec(&self, fds: &mut Vec<RawFd>) {
+        get_from_ancillary_data_by_vec(
+            &self.buffer[..self.length],
+            fds,
+            libc::SOL_SOCKET,
+            libc::SCM_RIGHTS,
+        );
+    }
+
+    /// Add file descriptors to the ancillary data.
+    ///
+    /// The function returns `true` if there was enough space in the buffer.
+    /// If there was not enough space then no file descriptors was appended.
+    /// Technically, that means this operation adds a control message with the level `SOL_SOCKET`
+    /// and type `SCM_RIGHTS`.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// #![feature(unix_socket_ancillary_data)]
+    /// use std::os::unix::net::{UnixStream, SocketAncillary};
+    /// use std::os::unix::io::AsRawFd;
+    /// use std::io::IoSliceMut;
+    ///
+    /// fn main() -> std::io::Result<()> {
+    ///     let sock = UnixStream::connect("/tmp/sock")?;
+    ///
+    ///     let mut ancillary_buffer = [0; 128];
+    ///     let mut ancillary = SocketAncillary::new(&mut ancillary_buffer[..]);
+    ///     ancillary.add_fds(&[sock.as_raw_fd()][..]);
+    ///
+    ///     let mut buf = [1; 8];
+    ///     let mut bufs = &mut [IoSliceMut::new(&mut buf[..])][..];
+    ///     sock.send_vectored_with_ancillary(bufs, &mut ancillary)?;
+    ///     Ok(())
+    /// }
+    /// ```
+    #[unstable(feature = "unix_socket_ancillary_data", issue = "none")]
+    pub fn add_fds(&mut self, fds: &[RawFd]) -> bool {
+        self.truncated = false;
+        add_to_ancillary_data(
+            &mut self.buffer,
+            &mut self.length,
+            fds,
+            libc::SOL_SOCKET,
+            libc::SCM_RIGHTS,
+        )
+    }
+
+    /// Get some credentials from this ancillary data into the specified buffer,
+    /// returning how many credentials were set.
+    ///
+    /// To receive a struct credentials, the `SO_PASSCRED` option must be enabled on the socket.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// #![feature(unix_socket_ancillary_data)]
+    /// use std::os::unix::net::{UnixStream, SocketAncillary};
+    /// use std::io::IoSliceMut;
+    ///
+    /// fn main() -> std::io::Result<()> {
+    ///     let sock = UnixStream::connect("/tmp/sock")?;
+    ///
+    ///     let mut cred = libc::ucred {
+    ///         pid: 0,
+    ///         uid: 0,
+    ///         gid: 0,
+    ///     };
+    ///     let mut ancillary_buffer = [0; 128];
+    ///     let mut ancillary = SocketAncillary::new(&mut ancillary_buffer[..]);
+    ///
+    ///     let mut buf = [1; 8];
+    ///     let mut bufs = &mut [IoSliceMut::new(&mut buf[..])][..];
+    ///     sock.recv_vectored_with_ancillary(bufs, &mut ancillary)?;
+    ///
+    ///     ancillary.get_creds(&mut [cred][..]);
+    ///     Ok(())
+    /// }
+    #[unstable(feature = "unix_socket_ancillary_data", issue = "none")]
+    pub fn get_creds(&self, creds: &mut [libc::ucred]) -> usize {
```

Comment:
We can't expose a struct directly from the libc crate - this interface should work with a wrapper type instead.

Reply:
Can you show me an example of the suggested wrapper struct?